pkg/stream/input.go

```diff
@@ -29,6 +29,9 @@ type AcknowledgeableInput interface {
 	// AckBatch does the same as calling Ack for every single message would, but it might use fewer calls to an external
 	// service.
 	AckBatch(ctx context.Context, msgs []*Message) error
+	// Nack does a negative acknowledgement of a message. If possible,
+	// prefer calling Nack with a batch as it is more efficient.
+	Nack(ctx context.Context, msg ...*Message) error
```

Comment:
Do I understand correctly, that calling Nack will be used to send messages into a dead letter queue?  
As the stream.consumer has already implemented a retry logic, I'm not sure if it makes sense to have something similar on the inputs.  
Do you plan to use the inputs independently or in conjunction with a consumer?  
If you just want to send messages into the dead letter queue instead of retrying them, we could think about improving the consumer retry handler.

Reply:
1. Yes, Nack will be used to send messages into a dead letter queue, since there are no per-message acks in kafka, but offsets, usually the way to do **non-blocking** retries is to dead-letter the message and have another consumer retry the message (from the dead-letter/retry topic), similar to approach explained here: https://eng.uber.com/reliable-reprocessing/

2. In conjunction with a consumer.

3. Plan would be similar to the approach in the article above, to achieve non-blocking retries by pushing to a retry/dead-letter topic and have another consumer (maybe parallel consumer) consume from there.

we could maybe provide multiple retry handler implementations? what do you think?

Reply:
Gosoline has already something in place to support this retry design. We're using it with our Kinesis consumers as Kinesis don't have built-in retry behaviour, too. You can configure it like this for a consumer:
```yaml
stream:
  consumer:
    default:
      input: kinesis-input
      retry:
        enabled: true
        type: sqs
        after: 5m
        max_attempts: 6
        queue_id: triggers-retry
```
If retry is enabled and the message isn't acked, it is send into the retry queue and gets reprocessed from there by the same consumer. After the max attempts it will end up in a DLQ. SQS is the default retry handler which provides this behaviour. If you want to stick with Kafka for retry, too, you can add a second implementation for the retry handler, but could also start with SQS first. Have a look into `gosoline/pkg/stream/retry.go` about this.

Reply:
thanks for the insights :) I will be working on this today.

Reply:
Pushed.